server/src/main/java/org/elasticsearch/monitor/fs/FsHealthService.java

```diff
@@ -0,0 +1,175 @@
+            Setting.Property.NodeScope);
+    public static final Setting<TimeValue> HEALTHY_TIMEOUT_SETTING =
+        Setting.timeSetting("monitor.fs.health.healthy_timeout", TimeValue.timeValueSeconds(1), TimeValue.timeValueMillis(1),
+            Setting.Property.NodeScope, Setting.Property.Dynamic);
+
+    public FsHealthService(Settings settings, ClusterSettings clusterSettings, ThreadPool threadPool, NodeEnvironment nodeEnv) {
+        this.threadPool = threadPool;
+        this.enabled = ENABLED_SETTING.get(settings);
+        this.refreshInterval = REFRESH_INTERVAL_SETTING.get(settings);
+        this.healthCheckTimeoutInterval = HEALTHY_TIMEOUT_SETTING.get(settings);
+        this.currentTimeMillisSupplier = threadPool::relativeTimeInMillis;
+        this.nodeEnv = nodeEnv;
+        this.pathHealthStats = new HashMap<>(1);
+        clusterSettings.addSettingsUpdateConsumer(HEALTHY_TIMEOUT_SETTING, this::setHealthCheckTimeoutInterval);
+        clusterSettings.addSettingsUpdateConsumer(ENABLED_SETTING, this::setEnabled);
+    }
+
+    @Override
+    protected void doStart() {
+        scheduledFuture = threadPool.scheduleWithFixedDelay(new FsHealthMonitor(), refreshInterval,
+                ThreadPool.Names.GENERIC);
+    }
+
+
+    @Override
+    protected void doStop() {
+        scheduledFuture.cancel();
+    }
+
+    @Override
+    protected void doClose() throws IOException {
+
+    }
+
+    public void setEnabled(boolean enabled) { this.enabled = enabled; }
+
+
+    public void setHealthCheckTimeoutInterval(TimeValue healthCheckTimeoutInterval) {
+        this.healthCheckTimeoutInterval = healthCheckTimeoutInterval;
+    }
+
+    @Override
+    public Status getHealth() {
+        if (enabled == false) {
+            return Status.UNKNOWN;
+        }
+        else if ((currentTimeMillisSupplier.getAsLong() - lastRunTimeMillis.get()) >
+            (refreshInterval.millis() + healthCheckTimeoutInterval.millis())) {
+            return Status.UNHEALTHY;
+        }
+        return pathHealthStats.entrySet().stream().anyMatch(map -> map.getValue() == Status.UNHEALTHY) ? Status.UNHEALTHY
+            : Status.HEALTHY;
+    }
+
+     class FsHealthMonitor implements Runnable {
+
+        static final String TEMP_FILE_NAME = ".es_temp_file";
+        private byte[] byteToWrite;
+
+        FsHealthMonitor(){
+            this.byteToWrite = new byte[20];
+        }
+        @Override
+        public void run() {
+            try {
+                if (enabled) {
+                    monitorFSHealth();
+                    logger.debug("Monitor for disk health ran successfully {}", pathHealthStats);
+                }
+            } catch (Exception e) {
+                logger.error("Exception monitoring disk health", e);
+            }
+        }
+
+        private void monitorFSHealth() {
+            if (checkInProgress.compareAndSet(false, true) == false) {
```

Comment:
Can that happen? I think `threadPool.scheduleWithFixedDelay` avoids this?

Reply:
My bad, I asked for this in https://github.com/elastic/elasticsearch/pull/52680#discussion_r383167077 it seems.

Reply:
:) I added it just an additional guarantee. I too think its not needed